src/docfx/lib/git/GitUtility.cs

```diff
@@ -213,6 +213,47 @@ public static void CheckMergeConflictMarker(string content, string file)
             }
         }
 
+        public static unsafe bool TryGetContentFromHistory(string repoPath, string filePath, string committish, out string content)
+        {
+            content = null;
+            if (git_repository_open(out var repo, repoPath) != 0)
+            {
+                return false;
+            }
+
+            if (git_revparse_single(out var commit, repo, committish) != 0)
+            {
+                git_repository_free(repo);
+                return false;
+            }
+
+            if (git_commit_tree(out var tree, commit) != 0)
+            {
+                git_repository_free(repo);
+                return false;
+            }
+
+            if (git_tree_entry_bypath(out var entry, tree, filePath) != 0)
+            {
+                git_repository_free(repo);
+                return false;
+            }
+
+            var obj = git_tree_entry_id(entry);
+            if (git_blob_lookup(out var blob, repo, obj) != 0)
+            {
+                git_tree_entry_free(entry);
+                git_repository_free(repo);
+                return false;
+            }
+
+            content = Marshal.PtrToStringUTF8(git_blob_rawcontent(blob));
```

Comment:
content may not be UTF8 encoded

Reply:
what's your suggestion, i refers LibGit2Sharp, they also by default using UTF8?

---
In reply to: [236529712](https://github.com/dotnet/docfx/pull/3677#discussion_r236529712) [](ancestors = 236529712)